### PR TITLE
[feature] add probablistic data structure bloomfilter

### PR DESCRIPTION
**Pull request by DengYiping**
Changes:
- add helper data structure bitmap implementation
- implement the bloomfilter using the bitmap. 
- necessary unit test for the bloomfilter
Note:
Bloom filter is completely probabilistic. There is a chance that the test case will fail, but it not likely to happen since the threshold I set in the test case is quite high

### CODE REVIEW

File: src/data-structures/bloomfilter.js [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+   * @param {Number} capacity the maximum capacity to maintain the given error rate
+   * @param {Number} errorRate the error rate expected under maximum capacity
+   */
+  exports.Bloomfilter = function(capacity, errorRate) {
+    errorRate = errorRate || 0.001; // default error rate
+    if (errorRate > 1 || errorRate < 0) {
+      throw 'the error rate range is outside of bound';
+    }
+    if (capacity < 0) {
+      throw 'the capacity cannot be negative';
```
Comment (mgechev) [inline src/data-structures/bloomfilter.js]: Let's use `Error` so the user can get a stack trace.

File: src/data-structures/bloomfilter.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+   * Create a bitmap with a given size
+   * Bit map is not resizeable
+   * @public
+   * @constructor
+   * @param {Number} size the size of the bitmap
+   */
+  exports.Bitmap = function(size) {
+    size = size || 1024;
+    if (size < 0) {
+      throw 'the size cannot be negative';
```
Comment (mgechev) [inline src/data-structures/bloomfilter.js]: Let's use `Error` so the user can get a stack trace. Also we can start with a capital letter.

File: src/data-structures/bloomfilter.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+   * @public
+   * @return {Boolean} true or false value of the bit
+   */
+  exports.Bitmap.prototype.get = function(index) {
+    // necessary boundary check
+    if (index < 0 || index > this.size) {
+      throw 'index out of bound';
+    }
+
+    var intOffset = index % 32; //calculate the offset within the int
```
Comment (mgechev) [inline src/data-structures/bloomfilter.js]: Would you add space after `//`

File: src/data-structures/bloomfilter.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+   * Construct a bloom filter by given the capacity and the error rate, default error rate is 0.001
+   * @public
+   * @constructor
+   * @param {Number} capacity the maximum capacity to maintain the given error rate
+   * @param {Number} errorRate the error rate expected under maximum capacity
+   */
+  exports.Bloomfilter = function(capacity, errorRate) {
+    errorRate = errorRate || 0.001; // default error rate
+    if (errorRate > 1 || errorRate < 0) {
+      throw 'the error rate range is outside of bound';
```
Comment (mgechev) [inline src/data-structures/bloomfilter.js]: Same here `new Error('The...')`

File: src/data-structures/bloomfilter.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  exports.Bitmap.prototype.getSize = function() {
+    return this.size;
+  };
+
+  /**
+   * get the bit at a particular index
+   * @public
+   * @return {Boolean} true or false value of the bit
+   */
+  exports.Bitmap.prototype.get = function(index) {
```
Comment (mgechev) [inline src/data-structures/bloomfilter.js]: Maybe we can call it `exists`? From `get` I'd expect to get a `1` or `0`, but this method returns a boolean.

**Comment by DengYiping** [comment]
@mgechev Thanks for the feedback. I will fix the implementation accordingly.

**Pull request merged**